src/java/org/apache/cassandra/index/sai/memory/FilteringKeyRangeIterator.java

```diff
@@ -17,30 +17,59 @@
  */
 package org.apache.cassandra.index.sai.memory;
 
+import java.io.IOException;
+import java.util.Iterator;
 import java.util.SortedSet;
 
+import com.google.common.collect.Iterators;
+import com.google.common.collect.PeekingIterator;
+
 import org.apache.cassandra.db.PartitionPosition;
 import org.apache.cassandra.dht.AbstractBounds;
 import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
 
 /**
- * A {@link KeyRangeIterator} that filters the returned {@PrimaryKey}s based on the provided keyRange
+ * A {@link RangeIterator} that filters the returned {@link PrimaryKey}s based on the provided keyRange
  */
-public class FilteringKeyRangeIterator extends KeyRangeIterator
+public class FilteringKeyRangeIterator extends RangeIterator
 {
     private final AbstractBounds<PartitionPosition> keyRange;
+    private final PeekingIterator<PrimaryKey> source;
 
     public FilteringKeyRangeIterator(SortedSet<PrimaryKey> keys, AbstractBounds<PartitionPosition> keyRange)
     {
-        super(keys);
+        super(keys.first(), keys.last(), keys.size());
         this.keyRange = keyRange;
+        this.source = Iterators.peekingIterator(keys.iterator());
     }
 
+    @Override
     protected PrimaryKey computeNext()
     {
-        PrimaryKey key = computeNextKey();
-        while (key != null && !keyRange.contains(key.partitionKey()))
-            key = computeNextKey();
-        return key == null ? endOfData() : key;
+        while (source.hasNext())
+        {
+            PrimaryKey key = source.next();
+            if (keyRange.contains(key.partitionKey()))
```

Comment:
As an aside, I wonder if we actually need this. We don't apply this same logic to sstable based indexes.

Reply:
Is it possible that here we have a key that is not in the range?
Where do we apply the filter upstream?

If we remove this check, is there any test failing ?

Reply:
Yes, we can have keys that are not in the requested range. We filter here:

https://github.com/datastax/cassandra/blob/1a16e6cf84f77c61269a8b4bae6c1f3d41a86a5c/src/java/org/apache/cassandra/index/sai/plan/StorageAttachedIndexSearcher.java#L240-L273

In the sstable indexes, we only use the range to skip sstable segment indexes that have zero overlap.

I spent some time looking into a solution earlier today, but I wasn't able to complete that work. I'll probably add to this PR in my morning tomorrow. Or, we can leave this as it is and merge it since it is already a large improvement.